### ISSUE TITLE
Add Ambari to the assembly

### ISSUE THREAD

**Issue by Nazarii-Melnyk** [rework-space-com/bigtop#11]
## Tasks
- [x] Choose LTS and stable version for Ambari
- [x] Add  Ambari artifact to build
- [x] Investigate Ambari
- [x] #15

**Comment by ya6lis** [comment]
Was choosen 2.7.5 version of Ambari because it was the last stable version that works on Ubuntu 22.04.

Official GitHub [repository](https://github.com/apache/ambari/tree/release-2.7.5)

**Comment by ya6lis** [comment]
For building Ambari was choose quide from official [documentation](https://cwiki.apache.org/confluence/display/AMBARI/Installation+Guide+for+Ambari+2.7.5)

The are some errors while running next comand:
``` mvn -B clean install jdeb:jdeb -DnewVersion=2.7.5.0.0 -DbuildNumber=5895e4ed6b30a2da8a90fee2403b6cab91d19972 -DskipTests -Dpython.ver="python >= 2.6" ```

Errors:
- 1 error
  ```
  [INFO] Reactor Summary:
  [INFO]
  [INFO] Ambari Main ....................................... SUCCESS [11:41.142s]
  [INFO] Apache Ambari Project POM ......................... SUCCESS [0.173s]
  [INFO] Ambari Web ........................................ FAILURE [33.222s]
  [INFO] Ambari Views ...................................... SKIPPED
  ...
  [ERROR]Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.4:yarn (yarn install) on project ambari-web: Failed 
  to run 
  task: 'yarn install --ignore-engines --pure-lockfile' failed. org.apache.commons.exec.ExecuteException: Process exited with an 
  error: 1
  ```
- Solution
  You need to change the version of node to yours in pom.xml (../apache-ambari-2.7.5-src/ambari-web/pom.xml)
  ```
  <nodeVersion>v4.5.0</nodeVersion> -> <nodeVersion>v12.22.9</nodeVersion>
  ```

- 2 error
  ```
  [INFO] Reactor Summary:
  [INFO] 
  [INFO] Ambari Main 2.7.5.0.0 .............................. SUCCESS [  2.326 s]
  [INFO] Apache Ambari Project POM 2.7.5.0.0 ................ SUCCESS [  0.042 s]
  [INFO] Ambari Web 2.7.5.0.0 ............................... SUCCESS [01:13 min]
  [INFO] Ambari Views 2.7.5.0.0 ............................. SUCCESS [  1.753 s]
  [INFO] Ambari Admin View 2.7.5.0.0 ........................ FAILURE [ 59.491 s]
  ...
  [ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (Bower install) on project ambari-admin: 
  Command execution failed. Process exited with an error: 1 (Exit value: 1) -> [Help 1]
  ```
- Solution
  You need to change the argument in pom.xml (../apache-ambari-2.7.5-src/ambari-admin/pom.xml)
  ```
  <argument>${basedir}/src/main/resources/ui/admin-web/node_modules/bower/bin/bower</argument> -> 
  <argument>bower</argument>
  ```

- 3 error
  ```
  [INFO] Reactor Summary:
  [INFO] 
  [INFO] Ambari Main 2.7.5.0.0 .............................. SUCCESS [  3.073 s]
  [INFO] Apache Ambari Project POM 2.7.5.0.0 ................ SUCCESS [  0.038 s]
  [INFO] Ambari Web 2.7.5.0.0 ............................... SUCCESS [ 53.607 s]
  [INFO] Ambari Views 2.7.5.0.0 ............................. SUCCESS [  0.942 s]
  [INFO] Ambari Admin View 2.7.5.0.0 ........................ SUCCESS [  6.139 s]
  [INFO] ambari-utility 1.0.0.0-SNAPSHOT .................... SUCCESS [  9.065 s]
  [INFO] ambari-metrics 2.7.5.0.0 ........................... SUCCESS [  0.296 s]
  [INFO] Ambari Metrics Common 2.7.5.0.0 .................... FAILURE [  0.448 s]
  ...
  [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.2:compile (default-compile) on project 
  ambari-metrics-common: Compilation failure: Compilation failure: 
  [ERROR] /home/ya6lis/apache-ambari-2.7.5-src/ambari-metrics/ambari-metrics- 
  common/src/main/java/org/apache/hadoop/metrics2/sink/timeline/TopNConfig.java:[24,33] package 
  javax.xml.bind.annotation does not exist
  [ERROR] /home/ya6lis/apache-ambari-2.7.5-src/ambari-metrics/ambari-metrics- 
 common/src/main/java/org/apache/hadoop/metrics2/sink/timeline/TopNConfig.java:[25,33] package javax.xml.bind.annotation 
  does not exist
  [ERROR] /home/ya6lis/apache-ambari-2.7.5-src/ambari-metrics/ambari-metrics- 
 common/src/main/java/org/apache/hadoop/metrics2/sink/timeline/TopNConfig.java:[26,33] package javax.xml.bind.annotation 
  does not exist
  ...
  ```
- Solution
  You need to add the dependecies in pom.xml (../apache-ambari-2.7.5-src/ambari-metrics/pom.xml)
  ```  <dependency>
    <groupId>javax.xml.bind</groupId>
    <artifactId>jaxb-api</artifactId>
    <version>2.3.0</version>
    </dependency>
  <dependency>
    <groupId>com.sun.xml.bind</groupId>
    <artifactId>jaxb-core</artifactId>
    <version>2.3.0</version>
  </dependency>
  <dependency>
    <groupId>com.sun.xml.bind</groupId>
    <artifactId>jaxb-impl</artifactId>
    <version>2.3.0</version>
  </dependency>
  ```

- 4 error
  ```
  [INFO] Reactor Summary:
  [INFO] 
  [INFO] Ambari Main 2.7.5.0.0 .............................. FAILURE [  2.896 s]
  ...
  [ERROR] Failed to execute goal org.apache.rat:apache-rat-plugin:0.12:check (default) on project ambari: Too many files with 
  unapproved license: 1 See RAT report in: /home/ya6lis/apache-ambari-2.7.5-src/target/rat.txt -> [Help 1]
  ```
- Solution
  Can`t find solution for this problem :(

**Comment by ya6lis** [comment]
Just try to installing a cluster with Ambari (with local VMs) using [guide](https://cwiki.apache.org/confluence/display/AMBARI/Quick+Start+Guide). 
- First problem becomes when cloning [ambari-vagrant](https://github.com/u39kun/ambari-vagrant) it only includes ubuntu from 12.4 to 18.4.
- The second problem is when running comman ``` wget -O /etc/yum.repos.d/ambari.repo http://public-repo-1.hortonworks.com/ambari/centos6/2.x/updates/2.5.1.0/ambari.repo ``` 
Gives the issue:
``` 
--2024-02-18 17:51:26--  http://public-repo-1.hortonworks.com/ambari/centos6/2.x/updates/2.5.1.0/ambari.repo
Resolving public-repo-1.hortonworks.com... 18.244.18.82, 18.244.18.113, 18.244.18.90, ...
Connecting to public-repo-1.hortonworks.com|18.244.18.82|:80... connected.
HTTP request sent, awaiting response... 403 Forbidden
2024-02-18 17:51:27 ERROR 403: Forbidden. 
```

**Comment by ya6lis** [comment]
@vitalyte 

To build the Ambari:
```
 ./gradlew allclean ambari-pkg-ind
 ```

To run the Ambari server from docker use the following command:
``` 
./gradlew docker-provisioner -Pconfig=config_ubuntu-22.04.yaml -POS=ubuntu-22.04 -Pprefix=3.2.1 -Pstack="ambari" 
```

But before that add ports to ```docker-compose.yml``` file, should look like this:
```
services:
    bigtop:
        image: ${DOCKER_IMAGE}
        command: /sbin/init
        domainname: bigtop.apache.org
        privileged: true
        mem_limit: ${MEM_LIMIT}
        volumes:
        - ../../:/bigtop-home
        - ./config/hiera.yaml:/etc/puppet/hiera.yaml
        - ./config/hieradata:/etc/puppet/hieradata
        - ./config/hosts:/etc/hosts
        ports:
            - "8080:8080"

networks:
    default:
        driver: bridge
        driver_opts:
            com.docker.network.driver.mtu: 1450
```

**Comment by vitalyte** [comment]
@ya6lis check these [files](https://github.com/apache/bigtop/tree/rel/3.1.1/bigtop-packages/src/common/ambari). 

1. Do you use it for bigtop run?
2. Why we need them?

**Comment by ya6lis** [comment]
@vitalyte 

1. We use it when creating a bigtop project (more precisely, when building ambari). They fixing some troubles that are necessary to build this version of bigtop.

**Comment by Nazarii-Melnyk** [comment]
@vitalyte 

As I see this patches bigtop uses to change something in the artifacts (in this case in ambari) source codes to fix some problems. Also I think they use this method because they want to change actual artifacts source code now in this versions. That allows them to not have to update the version of the particular artifact to use the required change.

Some of this patches can be very neccessary, For example file [patch5-AMBARI-25599.diff](https://github.com/apache/bigtop/blob/rel/3.1.1/bigtop-packages/src/common/ambari/patch5-AMBARI-25599.diff). This patch was created to solve issue [AMBARI25599](https://issues.apache.org/jira/browse/AMBARI-25599?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel) to update some references because previus will no longer available.

I also noticed that these patches often change and new ones are added. For example [that](https://github.com/apache/bigtop/tree/rel/3.1.1/bigtop-packages/src/common/ambari) how this directory looks on release 3.2.1. There was added new patch [patch13-AMBARI-25946.diff](https://github.com/TNTU-RS-internship/bigtop/blob/branch-3.2.1/bigtop-packages/src/common/ambari/patch13-AMBARI-25946.diff) that fixes CI/CD error for Ambari WebUI Tests.

Also I think not every patch are necessary. For example there are patch that fixes Ambari build on RHEL8. That are useless in our usage.
